storage/mysql/log_storage_test.go

```diff
@@ -244,12 +251,29 @@ func TestQueueDuplicateLeafFails(t *testing.T) {
 	defer tx.Close()
 
 	leaves := createTestLeaves(5, 10)
-	if err := tx.QueueLeaves(leaves, fakeQueueTime); err != nil {
+	if _, err := tx.QueueLeaves(leaves, fakeQueueTime); err != nil {
 		t.Fatalf("Failed to queue leaves: %v", err)
 	}
+	// Two overlapping sets of leaves:
+	//   [0] [1] [2] [3] [4]
+	//   10, 11, 12, 13, 14
+	//           12, 13, 14, 15, 16
+	//           [0] [1] [2] [3] [4]
+	// so first 3 are duplicates.
 	leaves2 := createTestLeaves(5, 12)
-	if err := tx.QueueLeaves(leaves2, fakeQueueTime); err == nil {
-		t.Fatal("Allowed duplicate leaves to be inserted")
+	existing, err := tx.QueueLeaves(leaves2, fakeQueueTime)
+	if err != nil {
+		t.Fatalf("Failed to re-queue leaves: %v", err)
+	}
+	for i := 0; i < 3; i++ {
+		if existing[i] == nil {
+			t.Errorf("entry[%d]=nil; want non-nil", i)
+		}
+	}
+	for i := 3; i < 5; i++ {
+		if existing[i] != nil {
+			t.Errorf("entry[%d]=%v; want nil", i, existing[i])
+		}
 	}
 	commit(tx, t)
```

Comment:
Aha, adding this test was helpful -- turns out the ordering of what's returned from MySQL isn't what's requested.  (Also, it just so happened that the 3 duplicate leaves did get returned in the right order, but the test failed when I extended it to a larger range!)